lib/Transforms/Utils/SplitModule.cpp

```diff
@@ -269,8 +269,13 @@ void llvm::SplitModule(
   // cloning it.
   for (unsigned I = 0; I < N; ++I) {
     ValueToValueMapTy VMap;
+    int HCCKernelIndex = 0;
     std::unique_ptr<Module> MPart(
         CloneModule(M.get(), VMap, [&](const GlobalValue *GV) {
+          if (const Function *F = dyn_cast<Function>(GV)) {
+            if (F->getCallingConv() == CallingConv::AMDGPU_KERNEL)
+              return ((HCCKernelIndex++) % N == I);
```

Comment:
It looks we are distributing the kernel functions among all partitions in a round robin fashion.  This may increase the number of local functions being externalized (see related "findPartitions" above...) and therefore, may increase the number of instances a local function being cloned into different partitions later on (and increase compilation cost).
It may be worthwhile look into a algorithm to minimize this effect (i.e. clustering kernel functions calling similar sets of local functions).
Also, it seems that the partitioning decision should belong to somewhere else and not at the point when the functions are being cloned.

Reply:
You're right, this is performing a round robin fashion with equal distribution. I was questioning this method before as well, and I think it might be better to not use this logic, since upstream llvm-split already performs an algorithm for best splitting. Do you think I should just get rid of this logic? The default version has better distribution and is more balanced, but we won't have equal distribution of kernel functions

Reply:
Is there any compilation cost analysis between the default and the round robin?

Reply:
The default will have more compilation cost analysis since it performs a Cluster Mapping of the Partitions to make. Default will call a findPartitions() function which will "Find partitions for module in the way that no locals need to be globalized. Try to balance pack those partitions into N files since this roughly equals thread balancing for the backend codegen step." This function does a bunch of analysis to create the Cluster mapping for the best distributed Balanced Queues. My round robin change might break the efficiency of that method.

Reply:
sorry my previous comment wasn't clear...
Have we done any cost comparison between the default and the round robin using real applications?  One concern I have is that only minimizing globalization of local functions may not be ideal since different partitions may have vastly sizes?

Reply:
I haven't done cost comparisons. I have looked at varying the number of partitions and threshold for splitting, but that can be better tweaked. I will try looking into cost comparison between default and round robin on rocFFT since that one has a bunch of 1000 kernel function files. Yes it's quite possible that there will be varying numbers in different partitions, I guess we will need to measure the timing to better understand it.

Reply:
@scchan I've looked into the times for building rocFFT using default splitModule function and my round-robin version. Here are the times:
Round Robin:
real    8m40.789s
user    38m27.032s
sys     1m13.574s
Default:
real    8m39.289s
user    38m56.203s
sys     1m6.181s
So to me it seems that there is no gain from using round robin or default method for kernel splitting. I think that the kernel splitting is done in such a short amount of time that it becomes negligible.

Reply:
Looking at the results, it's probably better to stick with the default splitter which takes into account the locals and externalization draw backs. I think make changes back to default.